lib/cpp/src/thrift/transport/TServerSocket.cpp

```diff
@@ -408,24 +407,21 @@ void TServerSocket::listen() {
     }
 
     struct sockaddr_un address;
+    // Store the zero-terminated path in address.sun_path
+    memset(&address, '\0', sizeof(address));
     address.sun_family = AF_UNIX;
-    memcpy(address.sun_path, path_.c_str(), len);
-
-    auto structlen = static_cast<socklen_t>(sizeof(address));
+    memcpy(address.sun_path, path_.c_str(), path_.size());
 
     if (!address.sun_path[0]) { // abstract namespace socket
-#ifdef __linux__
-      // sun_path is not null-terminated in this case and structlen determines its length
-      structlen -= sizeof(address.sun_path) - len;
-#else
+#ifndef __linux__
       GlobalOutput.perror("TSocket::open() Abstract Namespace Domain sockets only supported on linux: ", -99);
       throw TTransportException(TTransportException::NOT_OPEN,
                                 " Abstract Namespace Domain socket path not supported");
 #endif
     }
 
     do {
-      if (0 == ::bind(serverSocket_, (struct sockaddr*)&address, structlen)) {
+      if (0 == ::bind(serverSocket_, (struct sockaddr*)&address, static_cast<socklen_t>(sizeof(struct sockaddr_un)))) {
```

Comment:
I think, this will break the abstract unix socket type name.

Reply:
Can you elaborate further if you observed problems with this? I had the same problems as https://stackoverflow.com/a/11640827/7200132 and problems went away when switching to above solution. This was tested on Ubuntu 18.04 and CentOS 7.

Reply:
As the manual page says:

> The socket's address in this namespace is given by the
> additional bytes in sun_path that are covered by the specified
> length of the address structure.  (Null bytes in the name have no
> special significance.)

For abstract socket you need to pass the length of the string (without the null termination), plus sockaddr_un.sun_family size.

`static_cast<socklen_t>(sizeof(struct sockaddr_un)))` I think that will give the entire struct size

Reply:
Null termination, whether needed or not, is covered by the `memset`. Zero out the whole structure before use — that's the best way to go. So, null termination is *obviously correct* here.

Now, I think @deiv has a point too. The `addrlen` argument to `bind` may be incorrect. @emmenlau can you check if with this patch socket names in `/proc/net/unix` contain unexpected zero-bytes? Tests would still work this way I'm sure, but it's better to double-check this.

Reply:
Yes, as I say, the problem is the size not the null terminating character. But if you prefer to fill it with zeros, without needing it, go ahead !!